### PR TITLE
Avoid RegExp

### PR DESCRIPTION
**Pull request by tjconcept**
I was just playing around and this change gave a ~40% performance increase in latest Node.js.

### CODE REVIEW

**Comment by nickyout** [comment]
Cool. I'll check the performance improvement when I have time. Like, when I'm not at work. 

I intuitively knew exposing those two was a bad idea...

**Comment by tjconcept** [comment]
> Cool. I'll check the performance improvement when I have time. Like, when I'm not at work.

👍

Looking forward to see the results. How's your workflow for testing all those environments?

> I intuitively knew exposing those two was a bad idea...

I did wonder why, but figured you might have had some special use case yourself.

I do not fear breaking changes as long as:
- You use SemVer
- You have a good changelog describing an almost copy/paste-able upgrade path

File: index.js [hunk 1]
```diff
@@ -86,8 +67,32 @@ function sss(val) {
...
+		buffer = '',
+		code = 0,
+		i = 0;
+
+	for (; i < length; i++) {
+		code = str.charCodeAt(i);
+
+		if (code === 34) buffer += '\\"';
+		else if (code === 92) buffer += '\\\\';
+		else if (code > 31) buffer += String.fromCharCode(code);
```
Comment (tjconcept) [inline index.js:87]: This is intentionally moved here to optimise the common path, which I suspect is non-whitespace characters.

File: index.js [hunk 2]
```diff
@@ -86,8 +67,32 @@ function sss(val) {
...
+		else if (code > 15) buffer += "\\u00" + code.toString(16);
```
Comment (tjconcept) [inline index.js:88]: Same as above, cover as many cases as fast as possible.

**Comment by nickyout** [comment]
Apologies for the delay, I had some other things on my mind lately...

My testing workflow is... well, when I've made an alleged improvement on the code, I make sure to have installed zuul (`npm install -g zuul`) as that has been the easiest to use test runner I've used so far. I then run `./zuul-local.sh` and I'm given a test url for local testing. I open the url and check the console log for the difference in performance. Not a very automated workflow, but it was enough for me to signal a difference between substack's solution and mine. 

To extract a table of differences, I let travis run the test in several browsers using saucelabs. I download the log file, and use the script `makeTable.js` to get a table of differences out of it. In practice, travis or saucelabs gives unexpected errors and I have to run the test several times hoping it succeeds once. 

The local test indeed indicates a significant improvement in chrome:

```
Finished benchmarking: nickyout/fast-stable-stringify x 7,272 ops/sec ±1.37% (67 runs sampled) (cumulative string length: 106708052)
Finished benchmarking: substack/json-stable-stringify x 4,292 ops/sec ±1.78% (27 runs sampled) (cumulative string length: 69936031)
```

With `7272/4292 - 1 = 0.694` being a 69% improvement as opposed to my last check of 18% in recent chrome. 

Very cool. I'm looking forward to seeing the results in all browsers. Which is currently failing in travis. I'll see what I can do to fix that. 

As for the upgrade path... yeah, a copy/pasteable solution would be the proper thing to do. I suppose the solution is meant to be replaced like `str.replace(fss.stringSearch, fss.stringReplace)` to `fss.stringReplace(str)`, though it seems to me that the cleanest solution would be to actually create a new repository for this little sub component (`fast-stable-string-escape`?). I needed this little conversion somewhere else, and since I wanted high performance there as well, I exposed it. Ironic, really.

**Comment by nickyout** [comment]
I got the log of a decent run after trying to do a quick fix on the seemingly random testing errors. This is the table:

| Benchmark commit bd121510dd | nickyout/fast-stable-stringify | substack/json-stable-stringify | fastest* |
| --- | --- | --- | --- |
| safari 9 on Mac 10.11 | x 7,640 ops/sec | x 3,652 ops/sec | 109% |
| safari 5 on Windows 2008 | x 3,499 ops/sec | x 1,596 ops/sec | 119% |
| opera 12 on Windows 2003 | x 2,690 ops/sec | x 1,959 ops/sec | 37% |
| opera 11 on Windows 2003 | x 307 ops/sec | x 547 ops/sec | -43% |
| microsoftedge 13 on Windows | x 5,689 ops/sec | x 4,631 ops/sec | 22% |
| internet explorer 9 on Wind | x 5,064 ops/sec | x 3,005 ops/sec | 68% |
| internet explorer 11 on Win | x 5,674 ops/sec | x 4,638 ops/sec | 22% |
| internet explorer 10 on Win | x 6,192 ops/sec | x 3,684 ops/sec | 68% |
| firefox 20 on Windows 2008 | x 6,404 ops/sec | x 2,496 ops/sec | 156% |
| chrome 53 on Mac 10.11 | x 8,873 ops/sec | x 7,422 ops/sec | 19% |
| chrome 26 on Windows 2008 | x 8,356 ops/sec | x 3,993 ops/sec | 109% |
| android 5.1 on Linux | x 6,013 ops/sec | x 2,953 ops/sec | 103% |
| android 4.0 on Linux | x 9,184 ops/sec | x 6,822 ops/sec | 34% |

I should add the option to compare with last time. Either way, it suggests an ancient Opera version performs actually worse. Though finally it performs better at Android 4.0, haha. Chrome no longer has a big leap ahead...? Maybe it's because it was run on Mac. I'll check more later. 

The raw log I used is here: https://s3.amazonaws.com/archive.travis-ci.org/jobs/160911926/log.txt

**Comment by tjconcept** [comment]
Looking good 👍

I think you should benchmark against the master branch instead of Substack 🙂

**Pull request merged**